qiskit/passmanager/base_tasks.py

```diff
@@ -0,0 +1,227 @@
+        self,
+        status: WorkflowStatus,
+        run_state: RunState,
+    ):
+        """Update workflow status.
+
+        Args:
+            status: Status object to mutably update.
+            run_state: Completion status of current task.
+        """
+        status.previous_run = run_state
+        if run_state == RunState.SUCCESS:
+            status.count += 1
+            status.completed_passes.add(self)
+
+    @abstractmethod
+    def run(
+        self,
+        passmanager_ir: PassManagerIR,
+    ) -> PassManagerIR:
+        """Run optimization task.
+
+        Args:
+            passmanager_ir: Qiskit IR to optimize.
+
+        Returns:
+            Optimized Qiskit IR.
+        """
+        pass
+
+
+class BaseController(Task, ABC):
+    """Base class of controller.
+
+    A controller is built with a collection of pass manager tasks,
+    and a subclass provides a custom logic to choose next task to run.
+    Note a controller can be nested into another controller,
+    and controller itself doesn't provide any subroutine to modify the input IR.
+    """
+
+    def __init__(
+        self,
+        options: dict[str, Any] | None = None,
+    ):
+        """Create new flow controller.
+
+        Args:
+            options: Option for this flow controller.
+        """
+        self._options = options or {}
+
+    @abstractmethod
+    def iter_tasks(
+        self,
+        property_set: PropertySet,
+    ) -> Generator[Task]:
+        """A custom logic to choose a next task to run.
+
+        Controller subclass can consume the property set to build a proper task pipeline.
+        This indicates the order of task execution is only determined at running time.
+
+        Args:
+            property_set: A mutable data collection shared among all tasks.
+
+        Yields:
+            Next task to run.
+        """
+        pass
+
+    def execute(
+        self,
+        passmanager_ir: PassManagerIR,
+        status: WorkflowStatus | None = None,
+        property_set: PropertySet | None = None,
+        callback: Callable = None,
+    ) -> PassManagerIR:
+        # Overriding this method is not safe.
+        # Pass subclass must keep current implementation.
+        # Especially, task execution may break when method signature is modified.
+
+        if property_set is None:
+            property_set = PropertySet()
+        if status is None:
+            status = WorkflowStatus()
+
+        for next_task in self.iter_tasks(property_set=property_set):
+            try:
+                passmanager_ir = next_task.execute(
+                    passmanager_ir=passmanager_ir,
+                    status=status,
+                    property_set=property_set,
+                    callback=callback,
+                )
+            except TypeError as ex:
+                raise PassManagerError(
+                    f"{next_task.__class__} is not a valid pass for flow controller."
+                ) from ex
```

Comment:
In general I don't think we should do this kind of type-checking. If nothing else, there's a lot of reasons that a `TypeError` could be raised that _isn't_ because of `next_task` (some internal check of something in one of the arguments), and also ways in which `next_task` could not be a `Task` and _not_ raise `TypeError` (e.g. `().execute` raises `AttributeError`, not type error).

My main reason, though, is that it pushes things that should be developer-time assertions onto every single user's runtime; flow controllers should be responsible for ensuring that they're only accepting valid `Task`s into their pipelines at the point they're added, so we don't need to rescan everything on each use.

Reply:
This is preexisting behavior so I just copied. However removing this doesn't break any test, and I decided to drop this type check. Indeed I saw tons of confusing error message (i.e. not valid pass) when I was debugging the code. Done in [df0ee05](https://github.com/Qiskit/qiskit/pull/10127/commits/df0ee0530a8e19cf1a6664e59727a5fc8f460b10)